amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+        }
+        PyMem_Free(annotations);
+    }
+    ION_TYPE original_t = t;
+    IONCHECK(ion_reader_is_null(hreader, &is_null));
+    if (is_null) {
+        t = tid_NULL;
+    }
+    int ion_type = ION_TYPE_INT(t);
+
+    switch (ion_type) {
+        case tid_EOF_INT:
+            SUCCEED();
+        case tid_NULL_INT:
+        {
+            ION_TYPE null_type;
+            // Hack for ion-c issue https://github.com/amzn/ion-c/issues/223
+            if (original_t != tid_SYMBOL_INT) {
+                IONCHECK(ion_reader_read_null(hreader, &null_type));
+            }
+            else {
+                null_type = tid_SYMBOL_INT;
+            }
+
+            ion_type = ION_TYPE_INT(null_type);
+            py_value = Py_BuildValue(""); // INCREFs and returns Python None.
+            emit_bare_values = emit_bare_values && (ion_type == tid_NULL_INT);
+            ion_nature_constructor = _ionpynull_fromvalue;
+            break;
+        }
+        case tid_BOOL_INT:
+        {
+            BOOL bool_value;
+            IONCHECK(ion_reader_read_bool(hreader, &bool_value));
+            py_value = PyBool_FromLong(bool_value);
+            ion_nature_constructor = _ionpybool_fromvalue;
+            break;
+        }
+        case tid_INT_INT:
+        {
+            // TODO add ion-c API to return an int64 if possible, or an ION_INT if necessary
+            ION_INT ion_int_value;
+            IONCHECK(ion_int_init(&ion_int_value, hreader));
+            IONCHECK(ion_reader_read_ion_int(hreader, &ion_int_value));
+            SIZE int_char_len, int_char_written;
+            IONCHECK(ion_int_char_length(&ion_int_value, &int_char_len));
+            char* ion_int_str = (char*)PyMem_Malloc(int_char_len + 1); // Leave room for \0
+            err = ion_int_to_char(&ion_int_value, (BYTE*)ion_int_str, int_char_len, &int_char_written);
+            if (err) {
+                PyMem_Free(ion_int_str);
+                IONCHECK(err);
+            }
+            if (int_char_len < int_char_written) {
+                PyMem_Free(ion_int_str);
+                _FAILWITHMSG(IERR_BUFFER_TOO_SMALL, "Not enough space given to represent int as string.");
+            }
+            py_value = PyLong_FromString(ion_int_str, NULL, 10);
+            PyMem_Free(ion_int_str);
+
+            ion_nature_constructor = _ionpyint_fromvalue;
+            break;
+        }
+        case tid_FLOAT_INT:
+        {
+            double double_value;
+            IONCHECK(ion_reader_read_double(hreader, &double_value));
+            py_value = Py_BuildValue("d", double_value);
+            ion_nature_constructor = _ionpyfloat_fromvalue;
+            break;
+        }
+        case tid_DECIMAL_INT:
+        {
+            ION_DECIMAL decimal_value;
+            IONCHECK(ion_reader_read_ion_decimal(hreader, &decimal_value));
+            char dec_str[DEC_NUMBER_MAX_LEN]; // DECQUAD_String is only 43, which fails some tests.
```

Comment:
Default max decimal number is 43. I created a viable `DEC_NUMBER_MAX_LEN` = 300 to support more decimal digits.